### PR TITLE
Optional userland proxy

### PR DESCRIPTION
**Pull request by icecrime**
The `--userland-proxy` daemon flag makes it possible to rely on hairpin NAT and additional iptables routes instead of userland proxy for port publishing and inter-container communication.

Usage of the userland proxy remains the default as hairpin NAT is unsupported by older kernels.

To test _without_ userland-proxy:

```
DOCKER_USERLANDPROXY=false hack/make.sh binary test-integration-cli
```

Ping @phemmer for review!

### CODE REVIEW

File: daemon/networkdriver/bridge/driver.go [hunk 1]
```diff
@@ -230,17 +233,25 @@ func InitDriver(job *engine.Job) error {
 		}
 	}
 
+	if hairpinMode {
+		// Enable loopback adresses routing
+		sysPath := filepath.Join("/proc/sys/net/ipv4/conf", bridgeIface, "route_localnet")
```
Comment (erikh) [inline daemon/networkdriver/bridge/driver.go]: is there some reason you couldn't just test for this file and set the default that way?

File: daemon/networkdriver/bridge/driver.go [hunk 2]
```diff
@@ -230,17 +233,25 @@ func InitDriver(job *engine.Job) error {
...
+		if err := ioutil.WriteFile(sysPath, []byte{'1', '\n'}, 0644); err != nil {
+			logrus.Warnf("WARNING: unable to enable local routing for hairpin mode: %s\n", err)
```
Comment (calavera) [inline daemon/networkdriver/bridge/driver.go]: should we consider this an error?

File: docs/man/docker.1.md [hunk 3]
```diff
@@ -115,12 +118,13 @@ unix://[/path/to/socket] to use.
   Use TLS and verify the remote (daemon: verify client, client: verify daemon).
   Default is false.
 
+**--userland-proxy**=*true*|*false*
+    Rely on a userland proxy implementation for inter-container and outside-to-
+    container loopback communications. Default is true.
```
Comment (calavera) [inline docs/man/docker.1.md]: is this two lines for some reason?

**Comment by erikh** [comment]
I deleted my initial comment, but I'm wondering why this can't be tested at boot time and why that might be a bad idea over this solution?

**Comment by LK4D4** [comment]
All daemon tests failed :)
Also do we have someone with RHEL 6 to test if it works?
ping @mrunalp @vbatts

**Comment by icecrime** [comment]
Yes sorry it seems the second commit broke daemon restarts :-/

File: daemon/networkdriver/bridge/driver.go (hunk 2 above)
Comment (icecrime) [inline daemon/networkdriver/bridge/driver.go]: Good question :-/ It's the same kind of things we have [here](https://github.com/docker/docker/blob/06433cf812e26e468b9dcb6b0817f6597f1c397c/daemon/networkdriver/bridge/driver.go#L219), [here](https://github.com/docker/docker/blob/06433cf812e26e468b9dcb6b0817f6597f1c397c/daemon/networkdriver/bridge/driver.go#L225), or [here](https://github.com/docker/docker/blob/06433cf812e26e468b9dcb6b0817f6597f1c397c/daemon/networkdriver/bridge/driver.go#L228)... I'd rather keep it a warning for consistency, and maybe review the alert level of all in a followup PR.

**Comment by icecrime** [comment]
One change that I'd like to raise attention on is [this one](https://github.com/docker/docker/pull/12165/files#diff-6292b516b082a0f4416067b93cf8e725L43). Leaving this `! docker0` criteria in the iptables rules made it impossible for a container to connect to another container's exposed port through the host external address. I have no idea why this was part of the rule in the first place.

**Comment by phemmer** [comment]
> hairpin NAT is unsupported by older kernels

@LK4D4 [had mentioned](https://github.com/docker/docker/pull/10676#issuecomment-73952069) adding netlink support which would fix this issue. Never saw any follow up from that. Was the idea dropped?

**Comment by LK4D4** [comment]
@phemmer It was implemented. Dunno, maybe older kernels is 2.4 or 2.6.20.

**Comment by phemmer** [comment]
Yeah, I tested hairpin nat in RHEL6 and it worked aside from the whole `/sys` becoming read-only thing.
So since that's out of the way, I'd say we make disabling the proxy the default.

From reading through the code, all looks well. But I'll do a build tomorrow and play around with it for a more thorough review.

**Comment by LK4D4** [comment]
@phemmer Thank you very much.

**Comment by LK4D4** [comment]
@jfrazelle @calavera @crosbymichael @tiborvass @tianon @unclejack 
Share your thoughts on making it default. I'm +1, @icecrime -1 :)

**Comment by icecrime** [comment]
Haha :-) Let me give more details:
- I think it needs a bit of real world testing
- I think we need a followup PR to bind exposed ports during the **entire lifetime** of the container to ensure that:
   1. They are not already bound by something out of Docker (which would then be shadowed by the iptables rules)
   2. They cannot be bound later on by something outside of Docker (which would receive no error, even though the port is shadowed by iptables rules)

File: docs/sources/articles/networking.md [hunk 4]
```diff
@@ -446,6 +449,23 @@ address, you can edit your system-wide Docker server settings and add the
...
+> container. In such conflicting situation, Docker created iptables rules will
+take precedence and route to the container.
+
+The `--userland-proxy` parameter, true by default, provides a userland
+implementation for inter-container and outside-to-container communication. When
+disabled, Docker uses both an additional `MASQUERADE` iptable rule and the
+`net.ipv4.route_localnet` kernel parameter which allow the host machine to
+connect to a local container exposed port through the commonly used loopback
+address. Although this alternative is preferred for performance reason, its lack
+of support from older kernels makes the userland proxy implementation Docker's
```
Comment (LK4D4) [inline docs/sources/articles/networking.md]: This is not true now. At least for our supported kernels.

Comment (icecrime) [inline docs/sources/articles/networking.md]: Fixed

File: docs/sources/articles/networking.md [hunk 5]
```diff
@@ -446,6 +449,21 @@ address, you can edit your system-wide Docker server settings and add the
 option `--ip=IP_ADDRESS`.  Remember to restart your Docker server after
 editing this setting.
 
+> **Note**:
+> With hairpin NAT enabled (`--userland-proxy=false`), containers port exposure
+> is achieved purely through iptables rules, and no attempt to bind the exposed
+> port is ever made. This means that nothing prevents shadowing a previously
+> listening service outside of Docker through exposing the same port for a
+> container. In such conflicting situation, Docker created iptables rules will
+take precedence and route to the container.
```
Comment (LK4D4) [inline docs/sources/articles/networking.md]: Dunno if glitch - `>` seems missing.

File: daemon/networkdriver/portmapper/mapper.go [hunk 6]
```diff
@@ -121,11 +127,14 @@ func (pm *PortMapper) Map(container net.Addr, hostIP net.IP, hostPort int) (host
...
+		if err := proxy.Start(); err != nil {
+			if err := cleanup(); err != nil {
+				return nil, fmt.Errorf("Error during port allocation cleanup: %v", err)
+			}
+			return nil, err
 		}
-		return nil, err
+		m.userlandProxy = proxy
 	}
 	m.userlandProxy = proxy
```
Comment (LK4D4) [inline daemon/networkdriver/portmapper/mapper.go]: Seems redundant now

File: pkg/iptables/iptables.go [hunk 7]
```diff
@@ -90,8 +90,10 @@ func NewChain(name, bridge string, table Table) (*Chain, error) {
 		}
 		output := []string{
 			"-m", "addrtype",
-			"--dst-type", "LOCAL",
-			"!", "--dst", "127.0.0.0/8"}
+			"--dst-type", "LOCAL"}
+		if !hairpinMode {
+			output = append(output, "!", "--dst", "127.0.0.0/8")
```
Comment (LK4D4) [inline pkg/iptables/iptables.go]: Is this will break something if we put it without `hairpinMode`?

Comment (icecrime) [inline pkg/iptables/iptables.go]: ¯\_(ツ)_/¯

**Comment by cpuguy83** [comment]
We could always try to bind to the port before starting the container.

**Comment by LK4D4** [comment]
@cpuguy83 Yeah, that's the plan.

File: pkg/iptables/iptables.go (hunk 7 above)
Comment (LK4D4) [inline pkg/iptables/iptables.go]: We can try :) because it'll simplify code a lot.

File: docs/man/docker.1.md (hunk 3 above)
Comment (icecrime) [inline docs/man/docker.1.md]: Nope it's a mistake, thanks!

**Comment by LK4D4** [comment]
Okay, it hurts, but RHEL6 doesn't have netlink support for `IFLA_BRIDGE_MODE`.

**Comment by vbatts** [comment]
:-(
If we switch to this by default, then can the userland proxy flag stay there for use to build for RHEL6 without hairpin ?

**Comment by LK4D4** [comment]
@vbatts Of course flag will be there.
Also netlink support looks like supereasy to backport to 2.6 ;)

**Comment by mrunalp** [comment]
@LK4D4 I have pinged the kernel networking folks about this. I will let you'll know how it goes.

**Comment by tianon** [comment]
+1 for this userland proxy being disabled by default
+1 for including a flag to re-enable it
+1 for RHEL adding that flag to their default RPM-installed
"/etc/sysconfig/docker" if it turns out to be necessary there

**Comment by icecrime** [comment]
Totally agree with @tianon!

Do you mind if we start by getting this one in? I feel making `--userland-proxy=false` requires us to address the port binding issue first (which shouldn't be too hard).

**Comment by jfrazelle** [comment]
yes I like all the things \o/

**Comment by jfrazelle** [comment]
so can we move to code-review :)

File: daemon/networkdriver/bridge/driver.go (hunk 2 above)
Comment (calavera) [inline daemon/networkdriver/bridge/driver.go]: I think the `WARNING:` at the beginning of the message is superfluous and it can be removed, logrus already adds a `WARNING` to the log.

**Comment by calavera** [comment]
@icecrime this needs a rebase :cake:

**Comment by tiborvass** [comment]
@icecrime needs a rebase :S

**Comment by icecrime** [comment]
Rebased

**Comment by tiborvass** [comment]
code LGTM, tested with userlandproxy disabled on b2d.

**Comment by vbatts** [comment]
Do I need to retest on RHEL?

**Comment by icecrime** [comment]
@vbatts Thanks, I don't think that's necessary, there hasn't been any functional changes to it!

**Comment by LK4D4** [comment]
@icecrime Do we need to create issue about create stub listeners on each port anyway?

**Comment by icecrime** [comment]
@LK4D4 Yes you right that would be good.

File: daemon/networkdriver/bridge/driver.go (hunk 2 above)
Comment (LK4D4) [inline daemon/networkdriver/bridge/driver.go]: Yup, we removed `WARNING` from other places. Also `\n` is unnecessary.

Comment (icecrime) [inline daemon/networkdriver/bridge/driver.go]: Ok I'll change them all then.

**Comment by thaJeztah** [comment]
@LK4D4 if you create(d) that issue, could you add a reference to this PR? That will make it easier to find it back :)

**Comment by jfrazelle** [comment]
We should make sure to close some of the many userland proxy issues after merge ;)

**Comment by thaJeztah** [comment]
@jfrazelle ack. Will be a fun one to find them all, there may be some that don't actually *mention* the proxy, just "weird networking issue" :/

**Comment by LK4D4** [comment]
I didn't create anything btw :D

**Comment by jfrazelle** [comment]
LGTM

**Comment by calavera** [comment]
very nice. :dancer:  LGTM :dancer: . Moving to docs review.

File: docs/man/docker.1.md [hunk 8]
```diff
@@ -53,6 +53,9 @@ To see the man page for a command run **man docker <command>**.
 **-e**, **--exec-driver**=""
   Force Docker to use specific exec driver. Default is `native`.
 
+**--exec-opt**=[]
+  Set exec driver options. See EXEC DRIVER OPTIONS.
+
```
Comment (thaJeztah) [inline docs/man/docker.1.md]: Looks like an unrelated change. (I like it, but don't think it belongs in this PR)

File: docs/man/docker.1.md [hunk 9]
```diff
@@ -111,6 +114,9 @@ unix://[/path/to/socket] to use.
 **-s**, **--storage-driver**=""
   Force the Docker runtime to use a specific storage driver.
 
+**--selinux-enabled**=*true*|*false*
+  Enable selinux support. Default is false. SELinux does not presently support the BTRFS storage driver.
+
```
Comment (thaJeztah) [inline docs/man/docker.1.md]: Same here (unrelated change)

File: docs/man/docker.1.md [hunk 10]
```diff
@@ -121,15 +127,12 @@ unix://[/path/to/socket] to use.
   Use TLS and verify the remote (daemon: verify client, client: verify daemon).
   Default is false.
 
+**--userland-proxy**=*true*|*false*
+    Rely on a userland proxy implementation for inter-container and outside-to-container loopback communications. Default is true.
+
```
Comment (thaJeztah) [inline docs/man/docker.1.md]: I'd simplify the first part, e.g.

"Enable or disable the userland-proxy for inter-container and outside-to-container loopback communication. Default is true (enabled)."

Not sure about the "outside-to-container loopback communication" part, thinking of better wording for that :)

File: docs/sources/articles/networking.md [hunk 11]
```diff
@@ -452,6 +455,21 @@ address, you can edit your system-wide Docker server settings and add the
 option `--ip=IP_ADDRESS`.  Remember to restart your Docker server after
 editing this setting.
 
+> **Note**:
+> With hairpin NAT enabled (`--userland-proxy=false`), containers port exposure
+> is achieved purely through iptables rules, and no attempt to bind the exposed
+> port is ever made. This means that nothing prevents shadowing a previously
+> listening service outside of Docker through exposing the same port for a
+> container. In such conflicting situation, Docker created iptables rules will
+> take precedence and route to the container.
```
Comment (thaJeztah) [inline docs/sources/articles/networking.md]: If I understand correctly, this note will be removed after https://github.com/docker/docker/issues/13024 is implemented (which will in the same 1.7 release as this PR)

(If it is, I don't think we should invest too much time in perfectioning the text)

File: docs/sources/articles/networking.md [hunk 12]
```diff
@@ -452,6 +455,21 @@ address, you can edit your system-wide Docker server settings and add the
...
+
+The `--userland-proxy` parameter, true by default, provides a userland
+implementation for inter-container and outside-to-container communication. When
+disabled, Docker uses both an additional `MASQUERADE` iptable rule and the
+`net.ipv4.route_localnet` kernel parameter which allow the host machine to
+connect to a local container exposed port through the commonly used loopback
+address: this alternative is preferred for performance reason.
+
```
Comment (thaJeztah) [inline docs/sources/articles/networking.md]: I think these docs (including the note) may need some TLC, but it depends on how "permanent" these docs will be (also see my https://github.com/docker/docker/pull/12165/files#r29821743 above). 

It's a bit odd to mention the "alternative (non-default)" setting to actually be the preferred one. Do we know already if the default will be reversed for the 1.7 release? This could greatly simplify the docs here

File: docs/sources/reference/commandline/cli.md [hunk 13]
```diff
@@ -149,6 +149,7 @@ expect an integer, and they can only be specified once.
       --default-gateway-v6=""                Container default gateway IPv6 address
       --dns=[]                               DNS server to use
       --dns-search=[]                        DNS search domains to use
+      --default-ulimit=[]                    Set default ulimit settings for containers
```
Comment (thaJeztah) [inline docs/sources/reference/commandline/cli.md]: Unrelated change?

**Comment by thaJeztah** [comment]
I added some notes to the docs, but a bit in doubt how much time should be invested if the default will be reversed upon release and https://github.com/docker/docker/issues/13024 will also be implemented.

File: docs/man/docker.1.md (hunk 8 above)
Comment (icecrime) [inline docs/man/docker.1.md]: Alphabetical sort (I have OCD)

File: docs/man/docker.1.md (hunk 9 above)
Comment (icecrime) [inline docs/man/docker.1.md]: Alphabetical sort (I have OCD)

**Comment by icecrime** [comment]
Thanks for the review. I'm PTO until May 16th, so you may want to either wait for me, or merge and do a followup PR, sorry for this :-/

File: docs/sources/reference/commandline/cli.md (hunk 13 above)
Comment (icecrime) [inline docs/sources/reference/commandline/cli.md]: Alphabetical sort (I have OCD)

Comment (jfrazelle) [inline docs/sources/reference/commandline/cli.md]: We can fix whatever is left over ;) I say merge

File: docs/man/docker.1.md (hunk 8 above)
Comment (thaJeztah) [inline docs/man/docker.1.md]: heh, yes I noticed that (alphabetical); And I like it, just thought that should be a separate PR :)

**Comment by thaJeztah** [comment]
If we will indeed change the behavior (default) before the release (@jfrazelle ?), I personally see no reason to improve the docs at this point (which implies having to wait for @icecrime to return from PTO, or carry).

I'm fine with merging as-is, with the intent to improve before the actual release.

On that note; LGTM

ping @fredlf @jamtur01 @moxiegirl @svendowideit

**Comment by jfrazelle** [comment]
Going to merge because @icecrime is now on vacation and we can make  any follow up PRs after and I know he wont actually rest if we are still commenting here

**Comment by jfrazelle** [comment]
feel free to ping me on any modifications you want I am happy to oblidge

**Comment by Jaykah** [comment]
```docker run -d --userland-proxy=false myservice:latest```
```
flag provided but not defined: --userland-proxy
See 'docker run --help'.
```

```
# docker version
Client version: 1.7.1
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 786b29d
OS/Arch (client): linux/amd64
Server version: 1.7.1
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 786b29d
OS/Arch (server): linux/amd64
```

What am I missing?

**Comment by LK4D4** [comment]
@Jaykah it's daemon flag

**Comment by Jaykah** [comment]
@LK4D4 Sure, that's what I thought until I saw it in options:

```
#docker
Usage: docker [OPTIONS] COMMAND [arg...]

A self-sufficient runtime for linux containers.

Options:
...
  --tlsverify=false                    Use TLS and verify the remote
  --userland-proxy=true                Use userland proxy for loopback traffic
  -v, --version=false                  Print version information and quit
...

```

**Comment by LK4D4** [comment]
@Jaykah it's all daemon options. Not `docker run`.

**Comment by Jaykah** [comment]
@LK4D4 Oh, right. Got confused because I saw -d in there. Thanks.
